HistoryQueryManager/src/main/java/eu/neclab/ngsildbroker/historyquerymanager/service/HistoryQueryService.java

```diff
@@ -356,54 +373,106 @@ private Map<RemoteHost, Set<String>> getRemoteHostsForRetrieve(String tenant, St
+					regEntry.host().headers(), regEntry.host().cSourceId(), true, false, regEntry.regMode(),
+					regEntry.canDoZip(), regEntry.canDoIdQuery());
+
+			if(regEntry.eId() != null || regEntry.eIdp() != null){
+				for (String id:entityIds) {
+					if((regEntry.eId() != null && regEntry.eId().equals(id))
+						||  (regEntry.eIdp() != null && id.matches(regEntry.eIdp()))){
+						filteredIds.add(id);
+					}
 				}
-				if (regEntry.eRel() != null) {
-					// attribs.add(regEntry.eRel());
+			}
+			else{
+				filteredIds = entityIds;
+			}
+
+			if(regEntry.type() != null ){
+				for (String type:types) {
+					if( regEntry.type().equals(context.expandIri(type, false, true, null, null))) {
+						filteredTypes.add(type);
+					}
 				}
 			}
+			else{
+				filteredTypes=types;
+			}
+
+                queryMap.put("id",String.join(",",filteredIds));
+				queryMap.put("type",String.join(",",filteredTypes));
+
+            if (regEntry.eProp() != null || regEntry.eRel() != null) {
+                Set<String> uniqueAttrs = new HashSet<>( Arrays.asList(queryMap.getOrDefault("attrs","").split(",")));
+                uniqueAttrs.add(regEntry.eProp());
+                uniqueAttrs.add(regEntry.eRel());
+				uniqueAttrs.remove("");
+				uniqueAttrs.remove(null);
+                queryMap.put("attrs",String.join(",",uniqueAttrs));
+            }
+			result.put(remoteHost,queryMapToStr(queryMap));
 		}
-//		if (attrsQuery != null) {
-//			for (Entry<RemoteHost, Set<String>> entry : result.entrySet()) {
-//				Set<String> regAttrs = entry.getValue();
-//				if (regAttrs.isEmpty()) {
-//					regAttrs.addAll(attrsQuery.getAttrs());
-//				} else {
-//					regAttrs.retainAll(attrsQuery.getAttrs());
-//				}
-//			}
-//		}
 		return result;
 	}
+	private static Map<String, String> queryStrToMap(String queryString) {
+		Map<String, String> paramMap = new HashMap<>();
+
+		String[] paramPairs = queryString.split("&");
+		for (String paramPair : paramPairs) {
+			String[] keyValue = paramPair.split("=");
+			if (keyValue.length == 2) {
+				String key = keyValue[0];
+				String value = keyValue[1];
+				paramMap.put(key, value);
+			}
+		}
+
+		return paramMap;
+	}
+	public static String queryMapToStr(Map<String, String> paramMap) {
+		StringBuilder queryStringBuilder = new StringBuilder();
+		List<String> keyList = paramMap.keySet().stream().toList();
+        for (String key : keyList) {
+            if (paramMap.get(key).isEmpty()) {
+                paramMap.remove(key);
+            }
+        }
+
+		for (Map.Entry<String, String> entry : paramMap.entrySet()) {
+			if (!queryStringBuilder.isEmpty()) {
+				queryStringBuilder.append("&");
+			}
+			try {
+				String key = URLEncoder.encode(entry.getKey(), StandardCharsets.UTF_8);
+				String value = URLEncoder.encode(entry.getValue(), StandardCharsets.UTF_8);
+				queryStringBuilder.append(key).append("=").append(value);
+			} catch (Exception ignored) {}
```

Comment:
log the exception at  least. brought try catch with exception and no error handling is prone to bite us in the future.